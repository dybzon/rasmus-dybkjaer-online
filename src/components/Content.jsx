import * as React from 'react';
import styled from 'styled-components';

export const Content = props =>
(<StyledContent>
  {props.children}
</StyledContent>);

const StyledContent = styled.div`
  max-width: 800px;
  width: 100%;
  height: 100%;
  margin: auto;
  background-color: white;
  min-height: calc(100vh - 4rem - 8px);
`;