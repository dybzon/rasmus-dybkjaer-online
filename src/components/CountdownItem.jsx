import React from 'react';
import styled from 'styled-components';

export class CountdownItem extends React.PureComponent {
  render() {
    const { color, content, fontSize } = this.props;
    return (
      <Container color={color}>
        <Content fontSize={fontSize}>{content}</Content>
      </Container>
    );
  }
}

const Container = styled.div`
  height: 100%;
  flex-grow: 1;
  position: relative;
  background: ${props => props.color};
  margin: 0;
`;

const Content = styled.div`
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  font-size: ${props => props.fontSize || '1rem'};
  color: white;
`;
