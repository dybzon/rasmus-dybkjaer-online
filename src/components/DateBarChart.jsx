import styled from 'styled-components';
import React from 'react';
import { colors } from '../utilities';
import gift from '../images/gift.svg';

export class DateBarChart extends React.PureComponent {
  render() {
    const targetDate = new Date(this.props.targetDate);
    const now = new Date();
    const timeToTarget = targetDate.getTime() - now.getTime();
    const daysToTarget = Math.ceil(timeToTarget / (1000 * 60 * 60 * 24));
    const width = Math.ceil(daysToTarget / 365 * 100);
    const flexGrow = width / (100 - width);
    const textPart1 = daysToTarget;
    const textPart2 = `dag${daysToTarget > 1 && 'e'} til ${this.props.event}`;
    const textLeft = daysToTarget > 240 ? `${textPart1} ${textPart2}` : (daysToTarget < 10 ? '' : textPart1);
    const textRight = daysToTarget > 240 ? '' : (daysToTarget < 10 ? `${textPart1} ${textPart2}` : textPart2);
    return (
      <BarChartContainer>
        <BarChartFiller flexGrow={flexGrow} color={colors.lightGreen} isLeft>
          <TextContainerRight>{textLeft}</TextContainerRight>
          <GiftContainer />
        </BarChartFiller>
        <BarChartFiller flexGrow={1} color={colors.lightRose}>
          <TextContainerLeft>{textRight}</TextContainerLeft>
          <TextContainerLeftReverse>SÃ¥' der gaver til Rasmus!</TextContainerLeftReverse>
        </BarChartFiller>
      </BarChartContainer>
    );
  }
}

const BarChartContainer = styled.div`
  width: 100%;
  height: 30px;
  padding: 10px 0px;
  display: flex;
`;

const TextContainerRight = styled.div`
  position: absolute;
  right: 4px;
  top: 50%;
  transform: translate(0%, -50%);
  transition: opacity .5s linear;

  ${BarChartContainer}:hover & {
    opacity: 0;
  }    
`;

const TextContainerLeft = styled.div`
  position: absolute;
  left: 4px;
  top: 50%;
  transform: translate(0%, -50%);
  transition: opacity .5s linear;

  ${BarChartContainer}:hover & {
    opacity: 0;
  }
`;

const TextContainerLeftReverse = styled.div`
  position: absolute;
  left: 4px;
  top: 50%;
  transform: translate(0%, -50%);
  transition: opacity .5s linear;
  opacity: 0;

  ${BarChartContainer}:hover & {
    opacity: 1;
  }
`;

const GiftContainer = styled.div`
  background: url(${gift}) center no-repeat;
  position: absolute;
  top: 50%;
  left: 50%;
  transform: translate(-50%, -50%);
  height: 0%;
  width: 40px;
  transition: all .5s linear;

  ${BarChartContainer}:hover & {
    top: 50%;
    left: 0;
    transform: translate(-150%, -50%);
    height: 100%;
  }  
`;

const BarChartFiller = styled.div`
  position: relative;
  flex-grow: ${props => props.flexGrow};
  height: 100%;
  background: ${props => props.color};
  background-size: contain;
  border-radius: 4px;
  transition: flex-grow .5s;

  ${BarChartContainer}:hover & {
    ${props => props.isLeft && 'flex-grow: 0;'}
  }
`;
