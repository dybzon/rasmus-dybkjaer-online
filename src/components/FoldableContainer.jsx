import * as React from 'react';
import styled, { css } from 'styled-components';
import { colors } from '../utilities';

export class FoldableContainer extends React.Component {
  constructor(props) {
    super(props);
    this.state = { open: !!props.initialOpen }
  }

  render() {
    return (
      <Container {...this.state} {...this.props}>
        <ContainerHeader {...this.state} {...this.props} onClick={this.handleClick}>
          {this.props.header}
        </ContainerHeader>
        <ContainerContent {...this.state} maxHeight={this.props.maxContentHeight || 200}>
          {this.props.children}
        </ContainerContent>
      </Container>
    );
  }

  handleClick = () => {
    this.setState({ open: !this.state.open });
  }
}

const Container = styled.div`
  width: 100%;
  margin-bottom: 4px;
  color: ${colors.darkGrey};
  border-left: 1px solid transparent;
  border-top: 1px solid transparent;
  transition: border .2s ease-out;

  ${props => props.open && css`border-left: 1px solid ${(props.borderColor || colors.lightGreen)};`}
  :hover {
    ${props => !props.open && css`border-left: 1px solid ${props.borderColor || colors.lightGreen};`}
  }
`;

const ContainerHeader = styled.div`
  width: 100%;
  cursor: pointer;
  padding: 2px 6px;
  z-index: 10;
`;

const ContainerContent = styled.div`
  width: 100%;
  transition: all .2s ease-out;
  ${props => props.open 
    ? css`max-height: ${props.maxHeight}px;` 
    : css`max-height: 0px; overflow-y: hidden;`}
`;