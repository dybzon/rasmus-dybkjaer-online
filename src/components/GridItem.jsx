import * as React from 'react';
import styled, { css } from 'styled-components';

export const GridItem = props => 
(<StyledGridItem {...props}>
  {props.children}
</StyledGridItem>);

const StyledGridItem = styled.div`
  ${props => css`
    grid-column-start: ${props.gridColumnStart};
    grid-column-end: ${props.gridColumnEnd};
    grid-row-start: ${props.gridRowStart};
    grid-row-end: ${props.gridRowEnd};`}
`;
