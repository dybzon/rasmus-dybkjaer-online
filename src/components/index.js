export { ImageLink } from './ImageLink';
export { Content } from './Content';
export { ImageGrid } from './ImageGrid';
export { Menu } from './Menu';
export { FoldableContainer } from './FoldableContainer';
export { Wish } from './Wish';
export { Panorama } from './Panorama';
export { Background } from './Background';
export { DateBarChart } from './DateBarChart';
export { Countdown } from './Countdown';
export { Header1 } from './Header1';
export { InnerContent } from './InnerContent';
