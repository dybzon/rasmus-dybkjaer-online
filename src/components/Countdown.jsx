import React from 'react';
import styled from 'styled-components';
import { colors } from '../utilities';
import { CountdownItem } from './CountdownItem';
import { Header1 } from './Header1';
import { InnerContent } from './InnerContent';

export class Countdown extends React.Component {
  state = { now: new Date() };

  componentDidMount() {
    this.interval = setInterval(this.updateTime, 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  render() {
    const { event } = this.props;
    const targetDate = new Date(this.props.targetDate);
    const { now } = this.state;
    const timeToTarget = targetDate.getTime() - now.getTime();
    const msPerSecond = 1000;
    const msPerMinute = msPerSecond * 60;
    const msPerHour = msPerMinute * 60;
    const msPerDay = msPerHour * 24;
    const daysToTarget = Math.floor(timeToTarget / msPerDay);
    let msLeft = timeToTarget - daysToTarget * msPerDay;
    const hoursToTarget = Math.floor(msLeft / msPerHour);
    msLeft = msLeft - hoursToTarget * msPerHour;
    const minutesToTarget = Math.floor(msLeft / msPerMinute);
    msLeft = msLeft - minutesToTarget * msPerMinute;
    const secondsToTarget = Math.floor(msLeft / msPerSecond);
    return (
    <Background>
      <InnerContent>
        <Header1>Nedt√¶lling til {event}</Header1>
        <Container>
          <InnerContainer height={'75%'}>
            <CountdownItem color={colors.lightGreen} fontSize={70} content={daysToTarget} />
            <CountdownItem color={colors.lightGreen} fontSize={70} content={hoursToTarget} />
            <CountdownItem color={colors.lightGreen} fontSize={70} content={minutesToTarget} />
            <CountdownItem color={colors.lightGreen} fontSize={70} content={secondsToTarget} />
          </InnerContainer>
          <InnerContainer height={'25%'}>
            <CountdownItem content="dage" />
            <CountdownItem content="timer" />
            <CountdownItem content="minutter" />
            <CountdownItem content="sekunder" />
          </InnerContainer>
        </Container>
      </InnerContent>
    </Background>);
  }

  updateTime = () => {
    this.setState({ now: new Date() });
  }
}

const Background = styled.div`
  background: ${colors.lightRose};
`;

const Container = styled.div`
  height: 120px;
  width: 100%;
  display: flex;
  margin: 0 -10px;
  padding: 50px 0;
  justify-content: space-between;
  flex-direction: column;
`;

const InnerContainer = styled.div`
  width: 100%;
  height: ${props => props.height};
  display: flex;
  justify-content: center;
`;
