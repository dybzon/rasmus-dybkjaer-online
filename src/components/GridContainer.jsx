import * as React from 'react';
import styled, { css } from 'styled-components';
import { GridItem } from './GridItem';

// Define the number of grid columns for specific view port sizes
const defaultGridColumnIntervals = [
  { width: 500, columns: 1 },
  { width: 680, columns: 2 },
  { width: 780, columns: 3 },
  { width: 1200, columns: 4 },
  { width: 99999, columns: 5 },
];

/**
 * A container that takes any number of children and positions them in a grid
 * structure based on the width of the screen.
 * 
 * The width and gridColumnIntervals can be overwritten through props.
 */
export class GridContainer extends React.Component {
  constructor(props){
    super(props);
    this.state = { 
      gridColumnCount: 1, 
      width: this.props.width || window.innerWidth, 
      gridColumnIntervals: this.props.gridColumnIntervals || defaultGridColumnIntervals };
  }

  componentWillMount() {
    this.setState({ gridColumnCount: this.GridColumnCount });
  }

  componentDidMount() {
    // We may want to change the number of columns in the grid when the window is resized
    window.addEventListener('resize', this.handleResize);
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.handleResize);
  }

  render() {
    return (
      <Container {...this.state}>
        {this.GridItems}
      </Container>);
  }

  handleResize = () => {
    this.setState({ gridColumnCount: this.GridColumnCount });
  }

  get GridItems() {
    let counter = 1;
    const gridItems = [];
    const colCount = this.state.gridColumnCount;

    // Wrap each child in the GridContainer in a GridItem
    this.props.children.forEach((c, i) => {
      const rowStart = Math.ceil(counter / this.state.gridColumnCount);
      const colStart = (counter % colCount) === 0 ? colCount : (counter % colCount);
      counter++;
      gridItems.push(
        <GridItem 
          gridColumnStart={colStart} 
          gridColumnEnd={colStart+1} 
          gridRowStart={rowStart} 
          gridRowEnd={rowStart+1}
          key={i}>
            {c} 
          </GridItem>);
    });

    return gridItems;
  }

  get GridColumnCount() {
    // Get the number of columns matching the width.
    const width = this.state.width < window.innerWidth ? this.state.width : window.innerWidth;
    return this.state.gridColumnIntervals.find(g => g.width > width).columns;
  }
}

const Container = styled.div`
  display: grid;
  ${props => css`grid-template-columns: repeat(${props.gridColumnCount}, ${getGridItemWidth(props.gridColumnCount)});`} /* Consider changing column count through media query */
  grid-template-rows: auto;
  grid-gap: 10px;
  justify-content: center;`;

function getGridItemWidth(gridColumnCount) {
  return `${(100.0 / gridColumnCount)}%`;
}