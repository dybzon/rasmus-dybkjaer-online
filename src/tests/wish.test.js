import React from 'react';
import { Wish } from '../components';
import renderer from 'react-test-renderer';

test('Test snapshot of wish with price, without link', () => {
  const tree = renderer.create(
    <Wish name='Bits-sæt' category='Værktøj og teknik' price={70} />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Test snapshot of wish with price, with link', () => {
  const tree = renderer.create(
    <Wish name='Grimme puder' category='Til hjemmet' price={500} salesLink='http://puder.dk' />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

test('Test snapshot of wish without any price or link', () => {
  const tree = renderer.create(
    <Wish name='Sofabord' category='Til hjemmet' />,
  ).toJSON();
  expect(tree).toMatchSnapshot();
});