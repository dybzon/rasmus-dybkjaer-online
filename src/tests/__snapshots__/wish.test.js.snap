// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test snapshot of wish with price, with link 1`] = `
<div
  className="sc-kAzzGY crjdma"
  open={false}
>
  <div
    className="sc-chPdSV kBKTEg"
    onClick={[Function]}
    open={false}
  >
    Grimme puder
  </div>
  <div
    className="sc-kgoBCf kmIioY"
    open={false}
  >
    <div
      className="sc-kGXeez HSHAO"
    >
      Estimeret pris (DKK): 
      500
    </div>
    <div
      className="sc-kGXeez HSHAO"
    >
      <a
        href="http://puder.dk"
        target="blank"
      >
        Find den fx her...
      </a>
    </div>
  </div>
</div>
`;

exports[`Test snapshot of wish with price, without link 1`] = `
<div
  className="sc-kAzzGY crjdma"
  open={false}
>
  <div
    className="sc-chPdSV kBKTEg"
    onClick={[Function]}
    open={false}
  >
    Bits-s√¶t
  </div>
  <div
    className="sc-kgoBCf kmIioY"
    open={false}
  >
    <div
      className="sc-kGXeez HSHAO"
    >
      Estimeret pris (DKK): 
      70
    </div>
  </div>
</div>
`;

exports[`Test snapshot of wish without any price or link 1`] = `
<div
  className="sc-kAzzGY crjdma"
  open={false}
>
  <div
    className="sc-chPdSV kBKTEg"
    onClick={[Function]}
    open={false}
  >
    Sofabord
  </div>
  <div
    className="sc-kgoBCf kmIioY"
    open={false}
  />
</div>
`;
